File: src/main/java/com/guardanis/blank/tools/actions/ActionResolver.java
package com.guardanis.blank.tools.actions;

import java.util.ArrayList;
import java.util.List;

public class ActionResolver {

    public interface Resolvable {
        public boolean isResolutionRequired();
        public void resolve(Runnable successListener);
    }

    private List<Resolvable> resolvables = new ArrayList<Resolvable>();

    public ActionResolver add(Resolvable resolvable) {
        resolvables.add(resolvable);
        return this;
    }

    public void resolve(final Runnable successListener) {
        for(Resolvable resolvable : resolvables){
            if(resolvable.isResolutionRequired()){
                resolvable.resolve(() ->
                        resolve(successListener));

                return;
            }
        }

        successListener.run();
    }
}

File: src/main/java/com/guardanis/blank/tools/actions/ActionResolver.kt
package com.guardanis.blank.tools.actions

import android.content.Context
import java.util.ArrayList

class ActionResolver {

    private val resolvables = ArrayList<Resolvable>()

    interface Resolvable {
        fun isResolutionRequired(context: Context): Boolean
        fun resolve(context: Context, success: () -> Unit, fail: () -> Unit)
    }

    fun add(resolvable: Resolvable): ActionResolver {
        resolvables.add(resolvable)
        return this
    }

    fun resolve(context: Context, success: () -> Unit, fail: (() -> Unit)?) {
        for (resolvable in resolvables) {
            if (resolvable.isResolutionRequired(context)) {
                resolvable.resolve(context, { resolve(context, success, fail) }, fail ?: { })

                return
            }
        }

        success.invoke()
    }
}
